src/main/java/io/vertx/ext/mail/MailMain.java

```diff
@@ -0,0 +1,613 @@
+      return 500;
+    }
+    try {
+      return Integer.valueOf(message.substring(0, 3));
+    } catch (NumberFormatException n) {
+      return 500;
+    }
+  }
+
+  private boolean isStatusOk(String message) {
+    int statusCode = getStatusCode(message);
+    return statusCode >= 200 && statusCode < 400;
+  }
+
+//  private boolean isStatusFatal(String message) {
+//    return getStatusCode(message) >= 500;
+//  }
+
+  private boolean isStatusTemporary(String message) {
+    int statusCode = getStatusCode(message);
+    return statusCode >= 400 && statusCode < 500;
+  }
+
+  private void ehloCmd() {
+    write("EHLO " + getMyHostname(), result -> {
+      String message = result.result();
+      log.debug("EHLO result: " + message);
+      if (isStatusOk(message)) {
+        setCapabilities(message);
+
+        // fail if we are exceeding size as early as possible
+        createMailMessage();
+        if (capaSize > 0 && mailMessage.length() > capaSize) {
+          throwAsyncResult("message exceeds allowed size limit");
+        } else {
+          if (capaStartTLS && !ns.isSsl() && (config.getStarttls()==StarttlsOption.REQUIRED || config.getStarttls()==StarttlsOption.OPTIONAL)) {
+            // do not start TLS if we are connected with SSL
+            // or are already in TLS
+            startTLSCmd();
+          } else {
+            if (!ns.isSsl() && config.getStarttls()==StarttlsOption.REQUIRED) {
+              log.warn("STARTTLS required but not supported by server");
+              throwAsyncResult("STARTTLS required but not supported by server");
+            } else {
+              if (config.getLogin() != LoginOption.DISABLED && config.getUsername() != null && config.getPassword() != null && !capaAuth.isEmpty()) {
+                authCmd();
+              } else {
+                if (config.getLogin() == LoginOption.REQUIRED) {
+                  if (config.getUsername() != null && config.getPassword() != null) {
+                    throwAsyncResult("login is required, but no AUTH methods available. You may need do to STARTTLS");
+                  } else {
+                    throwAsyncResult("login is required, but no credentials supplied");
+                  }
+                } else {
+                  mailFromCmd();
+                }
+              }
+            }
+          }
+        }
+      } else {
+        // if EHLO fails, assume we have to do HELO
+        if (isStatusTemporary(message)) {
+          heloCmd();
+        } else {
+          throwAsyncResult("EHLO failed with " + message);
+        }
+      }
+    });
+  }
+
+  /**
+   * @param message
+   */
+  private void setCapabilities(String message) {
+    List<String> capabilities = parseEhlo(message);
+    for (String c : capabilities) {
+      if (c.equals("STARTTLS")) {
+        capaStartTLS = true;
+      }
+      if (c.startsWith("AUTH ")) {
+        capaAuth = new HashSet<String>(Arrays.asList(c.substring(5).split(" ")));
```

Comment:
Avoid to use split as it involves creating/parsing a regex. Rather use a static method with indexOf(' ') split.